velox/functions/prestosql/aggregates/PrestoHasher.cpp

```diff
@@ -0,0 +1,182 @@
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::VARCHAR>(vector_size_t row) {
+  return hashStringView(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::VARBINARY>(vector_size_t row) {
+  return hashStringView(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::REAL>(vector_size_t row) {
+  return hashInteger(vector_->valueAt<int32_t>(row));
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::TIMESTAMP>(vector_size_t row) {
+  return hashInteger<int64_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::BOOLEAN>(vector_size_t row) {
+  auto value = vector_->valueAt<bool>(row);
+  return value ? 1231 : 1237;
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::ARRAY>(vector_size_t row) {
+  auto baseArray = vector_->base()->as<velox::ArrayVector>();
+  auto index = vector_->index(row);
+  auto size = baseArray->sizeAt(index);
+  auto offset = baseArray->offsetAt(index);
+
+  if (children_.empty()) {
+    children_.reserve(1);
+    children_.push_back(std::make_unique<PrestoHasher>(*baseArray->elements()));
+  }
+
+  int64_t hash = 0;
+  for (int i = 0; i < size; i++) {
+    hash = 31 * hash + children_[0]->computeHash(offset + i);
+  }
+
+  return hash;
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::MAP>(vector_size_t row) {
+  auto baseMap = vector_->base()->as<MapVector>();
+  auto index = vector_->index(row);
+  auto size = baseMap->sizeAt(index);
+  auto offset = baseMap->offsetAt(index);
+
+  if (children_.empty()) {
+    children_.reserve(2);
+    // Decode key
+    children_.push_back(std::make_unique<PrestoHasher>(*baseMap->mapKeys()));
+
+    // Decode values
+    children_.push_back(std::make_unique<PrestoHasher>(*baseMap->mapValues()));
+  }
+
+  VELOX_CHECK_EQ(children_.size(), 2)
+
+  auto keyIndices = children_[0]->indices();
+  auto valueIndices = children_[1]->indices();
+
+  int64_t hash = 0;
+  for (int i = 0; i < size; i++) {
+    auto keyHash = children_[0]->computeHash(keyIndices[offset] + i);
+    auto valueHash = children_[1]->computeHash(valueIndices[offset] + i);
+    hash = keyHash ^ valueHash;
+  }
+  return hash;
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::ROW>(vector_size_t row) {
+  auto baseRow = vector_->base()->as<RowVector>();
+
+  if (children_.empty()) {
+    children_.reserve(baseRow->childrenSize());
+    for (int i = 0; i < baseRow->childrenSize(); i++) {
+      children_.push_back(std::make_unique<PrestoHasher>(*baseRow->childAt(i)));
+    }
+  }
+
+  int64_t hash = 1;
+  for (int i = 0; i < baseRow->childrenSize(); i++) {
+    hash = 31 * hash + children_[i]->computeHash(row);
+  }
+
+  return hash;
+}
+
+int64_t PrestoHasher::computeHash(vector_size_t row) {
+  if (vector_->isNullAt(row)) {
+    return 0;
+  }
+
+  auto kind = vector_->base()->typeKind();
+  return VELOX_DYNAMIC_TYPE_DISPATCH(computeHash, kind, row);
```

Comment:
Type dispatch for every row tends to be slow. Consider adding a benchmark to compare hashing a flat vector of integers using PrestoHasher and hand-written loop.